38377 direct scheduling reason for appointment comment not in payload

vbahinwillit opened the pull request:
## Description
This PR sends the reason for appointment information in the 'reasonCode' field

## Original issue(s)
department-of-veterans-affairs/va.gov-team#38377


## Testing done
Unit testing

## Screenshots
N/A

## Acceptance criteria
- [ ] All test should pass

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

vbahinwillit committed:
modified to include comment when creating an appointment

vbahinwillit committed:
modified to include comment when creating an appointment

vbahinwillit committed:
modified to include comment when creating an appointment

vbahinwillit committed:
Merge branch 'main' into 38377-direct-scheduling-reason-for-appointment-comment-not-in-payload

vbahinwillit committed:
Merge branch 'main' into 38377-direct-scheduling-reason-for-appointment-comment-not-in-payload

vbahinwillit committed:
modified to send reasonCode instead of comment

vbahinwillit committed:
Merge branch 'main' into 38377-direct-scheduling-reason-for-appointment-comment-not-in-payload

jenniemc reviewed (changes requested):
New appointment request and new direct schedule use the same POST call for V2, and since appointment request comment is correctly passed, we could mimic the same logic/code for the direct schedule comment.

jenniemc commented on src/applications/vaos/new-appointment/redux/helpers/formSubmitTransformers.v2.js:
```diff
@@ -155,12 +155,23 @@ export function transformFormToVAOSVARequest(state) {
-  return `${label}: ${data.reasonAdditionalInfo}`;
+  return {
+    coding: code ? [{ code }] : [],
+    text: data.reasonAdditionalInfo,
```
we return `text` key value pair only if user selects "My reason isn't listed". Plus, the `text` field may be limited to 100 characters. Not quite sure how to code it in ternary expression or maybe a simple if statement will be better?
```suggestion
    text: !code? data.reasonAdditionalInfo: null,
```

jenniemc commented on src/applications/vaos/new-appointment/redux/helpers/formSubmitTransformers.v2.js:
```diff
@@ -181,7 +192,8 @@ export function transformFormToVAOSAppointment(state) {
+    reasonCode: getReasonCode(data),
     // removing this for now, it's preventing QA from testing, will re-introduce when the team figures out how we're handling the comment field
-    comment: getUserMessage(data),
+    // comment: getUserMessage(data),
```
Since this is where the comment key:value pair is transformed for appointment request, we can add it for direct schedule appointment
```suggestion
    comment: data.reasonAdditionalInfo,
```

vbahinwillit replied:
Did not change since this change is in the 'transformFormToVAOSApppointment' (direct scheduling) function.

jenniemc commented on src/applications/vaos/tests/new-appointment/components/ReviewPage/index.direct.unit.spec.jsx:
```diff
@@ -483,7 +484,14 @@ describe('VAOS <ReviewPage> direct scheduling with v2 api', () => {
+            code: 'Follow-up/Routine',
+          },
+        ],
+        text: 'I need an appt',
```
delete `text:` It is not required for radio options besides "My reason isn't listed"

vbahinwillit replied:
Done

jenniemc commented on src/applications/vaos/tests/new-appointment/redux/helpers/formSubmitTransformers.v2.unit.spec.js:
```diff
@@ -113,7 +114,198 @@ describe('VAOS V2 data transformation', () => {
+              code: 'Follow-up/Routine',
+            },
+          ],
+          text: 'asdfasdf',
```
delete `text:` It is not required for radio options besides "My reason isn't listed"

vbahinwillit replied:
Done

jenniemc commented on src/applications/vaos/tests/new-appointment/redux/helpers/formSubmitTransformers.v2.unit.spec.js:
```diff
@@ -113,7 +114,198 @@ describe('VAOS V2 data transformation', () => {
+            phoneNumber: '2234567890',
+            email: 'joeblow@gmail.com',
+            reasonForAppointment: 'routine-follow-up',
+            reasonAdditionalInfo: 'I need an appt',
```
delete `text:` It is not required for radio options besides "My reason isn't listed"

jenniemc commented on src/applications/vaos/tests/new-appointment/redux/helpers/formSubmitTransformers.v2.unit.spec.js:
```diff
@@ -113,7 +114,198 @@ describe('VAOS V2 data transformation', () => {
+              code: 'Follow-up/Routine',
+            },
+          ],
+          text: 'I need an appt',
```
delete `text:` It is not required for radio options besides "My reason isn't listed"

vbahinwillit replied:
Done

vbahinwillit committed:
modified per code review

vbahinwillit committed:
modified to use v2 constant
fixed PURPOSE_TEXT_V2 error

vbahinwillit committed:
fix unit test

vbahinwillit committed:
fix unit test

jenniemc reviewed (commented):
This review instance is sending the reasonCode in the payload. Just need to have Marcy check if the two newly created direct schedule appointments have its reason for appointment recorded in the vistA for cecil.morgan@id.me: 

- id: 108070, Audiology on 10/13/22, reasonCode.text =“testing for the comment section of free text when selecting the “other” radio option to check where ”

- id: 108069, Primary Care on 7/15/22, reasonCode.coding[0] = “Routine Follow-up”

vbahinwillit commented:
Looks like it's working! Here are the results from Marcy:

```
Here is the data on the 7/15 appt:
  Name: MORGAN,CECIL MATTHEW           Clinic: CHY C&P KILPATRICK
             ID: 796-10-6146                 Date/Time: JUL 15, 2022@14:20
         Status: FUTURE
Purpose of Vst.: SCHEDULED
 Length of Appt: 20                          Appt Type: REGULAR
            Lab:                          Elig of Appt: SC LESS THAN 50%
          X-ray:                              Overbook: NO
            EKG:                       Collateral Appt: NO
          Other: Routine Follow-up
```
```
10/13/2022:
Name: MORGAN,CECIL MATTHEW           Clinic: FTC C&P AUDIO BEV
             ID: 796-10-6146                 Date/Time: OCT 13, 2022@09:00
         Status: FUTURE
Purpose of Vst.: SCHEDULED
 Length of Appt: 60                          Appt Type: REGULAR
            Lab:                          Elig of Appt: SC LESS THAN 50%
          X-ray:                              Overbook: NO
            EKG:                       Collateral Appt: NO
          Other:testing for the comment section of free text when selecting the
"other" radio option to check where 
```

vbahinwillit committed:
changed reason for appt. max char count (V2 only).
default char limit for CC appointment is 250
default char limit for both VA appointment types is 100

vbahinwillit committed:
changed reason for appt. max char count (V2 only).
default char limit for CC appointment is 250
default char limit for both VA appointment types is 100

vbahinwillit committed:
Merge branch 'main' into 38377-direct-scheduling-reason-for-appointment-comment-not-in-payload

jenniemc reviewed (changes requested):
As discussed, the VA appointment details page is missing the below.  The [content copy doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/health-care/appointments/va-online-scheduling/content/copy-docs/appointment-details.md#va-booked-future)  references V0 data fields, please adjust for v2 in ConfirmedAppointmentDetailsPage/VAInstructions.jsx 
```
You shared these details about your concern
{Reason}: {Booking note}
```

jenniemc commented on src/applications/vaos/new-appointment/redux/helpers/formSubmitTransformers.v2.js:
```diff
@@ -181,7 +181,8 @@ export function transformFormToVAOSAppointment(state) {
       desiredDate: `${data.preferredDate}T00:00:00+00:00`,
     },
     locationId: data.vaFacility,
+    reasonCode: getReasonCode(data),
```
Type of care is missing in the VA appointment details page. 
Add `serviceType: typeOfCare.idV2,` to pass in type of care

vbahinwillit replied:
I think we need to create another ticket to update all mock data to match the current swagger docs.

vbahinwillit commented on src/applications/vaos/new-appointment/redux/helpers/formSubmitTransformers.v2.js:
```diff
@@ -155,12 +151,29 @@ export function transformFormToVAOSVARequest(state) {
+    // If the veteran selects the "other reason" radio button and enters a free text reason
+    // then you will want to leave reasonCode.coding[] as null/empty, and then send the free
+    // text reason in reasonCode.text :
+    text: code ? undefined : data.reasonAdditionalInfo.slice(0, 100),
```
I was decided that it should be truncated to 100 characters.

jenniemc commented:
Using review instance and test user Cathedris.Mitulski@id.me, 
the VA appointment request and VA direct schedule payload behaves as expected with passing the reasonCode objects. The only concern is when displaying the detailed pages, it does not match what's in v0. 

**VA Direct Schedule using V0 (var resources) and Ruben Moreno login**
![image](https://user-images.githubusercontent.com/54327023/175723228-e8215d62-2989-415e-b286-ff20b437ac62.png)
**VA Direct Schedule using V2 (vaos service)**
![image](https://user-images.githubusercontent.com/54327023/175723946-81104600-1537-455d-bb67-be1e97350ca9.png)

**VA Request Appt using V0 (var resources) and Ruben Moreno login**
![image](https://user-images.githubusercontent.com/54327023/175731660-dd3db76a-916c-47b3-8eb3-f35f5ee46372.png)

**VA Request Appt using V2 (vaos service)**
![image](https://user-images.githubusercontent.com/54327023/175733385-7132d125-cf1f-4ec3-8abb-3b2b804e795a.png)

vbahinwillit committed:
fix reasonCode data mapping

vbahinwillit committed:
fix unit test

vbahinwillit committed:
Merge branch 'main' into 38377-direct-scheduling-reason-for-appointment-comment-not-in-payload

vbahinwillit committed:
fix unit test

jenniemc commented:
@vbahinwillit VA appointment request in the review instance using Noah James as test user. When selecting "other" radio option and entering free text works as expected. VA appt request selecting "routine" and entering free text only displays the free text. It does not show the reason code value.
![Screen Shot 2022-06-28 at 10 59 17 AM](https://user-images.githubusercontent.com/54327023/176254550-97400a2f-b329-4fc9-a4da-341c49e15089.png)

vbahinwillit committed:
fix to use serviceName instead of short

vbahinwillit committed:
fix to use serviceName instead of short

vbahinwillit committed:
fix to use serviceName instead of short

vbahinwillit committed:
fix to concatenate reason code and text

vbahinwillit committed:
fix mapping
